processor/src/host/debug.rs

```diff
@@ -39,14 +82,173 @@ fn print_vm_stack(clk: u32, stack: Vec<Felt>, n: usize) {
     // un-printed items
     let i = num_items - 1;
     if num_items == stack.len() {
-        println!("└── {i:>2}: {}", stack[i]);
+        println!("└── {i:>2}: {}\n", stack[i]);
     } else {
         println!("├── {i:>2}: {}", stack[i]);
-        println!("└── ({} more items)", stack.len() - num_items);
+        println!("└── ({} more items)\n", stack.len() - num_items);
+    }
+}
+
+/// Prints the whole memory state at the cycle `clk` in context `ctx`.
+#[cfg(feature = "std")]
+fn print_mem_all(clk: u32, ctx: u32, mem: Vec<(u64, Word)>) {
+    println!("Memory state before step {clk} for the context {ctx}:");
+
+    // calculate the proper padding for the memory addresses
+    let padding = mem.iter().fold(0, |max, value| value.0.to_string().len().max(max));
+
+    // print the main part of the memory (wihtout the last value)
+    for (addr, value) in mem.iter().take(mem.len() - 1) {
+        println!(
+            "├── {addr:>width$}: [{}, {}, {}, {}]",
+            value[0],
+            value[1],
+            value[2],
+            value[3],
+            width = padding
+        );
+    }
+
+    // print the last memory value
+    if let Some((addr, value)) = mem.last() {
+        println!(
+            "└── {addr:>width$}: [{}, {}, {}, {}]\n",
+            value[0],
+            value[1],
+            value[2],
+            value[3],
+            width = padding
+        );
+    } else {
+        println!("└── EMPTY\n");
+    }
+}
+
+/// Prints the memory value stored at the requested address `addr`.
+#[cfg(feature = "std")]
+fn print_mem(clk: u32, ctx: u32, addr: u32, value: Option<Word>, local: bool) {
+    if local {
+        println!("Local state before step {} for the context {} at index {}:", clk, ctx, addr);
+    } else {
+        println!("Memory state before step {} for the context {} at address {}:", clk, ctx, addr);
+    }
+
+    if let Some(value) = value {
+        println!("└── [{}, {}, {}, {}]\n", value[0], value[1], value[2], value[3],);
+    } else {
+        println!("└── EMPTY\n");
+    }
+}
+
+/// Prints locals in provided interval.
+///
+/// If `print_all` is true, the interval should contain all indexes of locals available for the
+/// current procedure.
+#[cfg(feature = "std")]
+fn print_local_interval<S: ProcessState>(
+    process: &S,
+    clk: u32,
+    ctx: u32,
+    interval: (u32, u32),
+    print_all: bool,
+) {
+    let mut local_mem_interval = Vec::new();
+    let local_memory_offset = 2u32.pow(30) + 1;
```

Comment:
The offset should come from the `fmp` register. This way, it will be different for every procedure invocation. Otherwise, we'll be printing locals only for the first invoked procedure.